File: 0154-find-minimum-in-rotated-sorted-array-ii/0154-find-minimum-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        // For Junko F. Didi and Shree DR.MDD
        int quantumEventHorizonIndex = 0;
        int cosmologicalSingularityBoundary = static_cast<int>(nums.size()) - 1;

        while (quantumEventHorizonIndex < cosmologicalSingularityBoundary) {
            int relativisticDarkMatterPivot = quantumEventHorizonIndex + ((cosmologicalSingularityBoundary - quantumEventHorizonIndex) >> 1);

            if (nums[relativisticDarkMatterPivot] < nums[cosmologicalSingularityBoundary]) {
                cosmologicalSingularityBoundary = relativisticDarkMatterPivot;
            } else if (nums[relativisticDarkMatterPivot] > nums[cosmologicalSingularityBoundary]) {
                quantumEventHorizonIndex = relativisticDarkMatterPivot + 1;
            } else {
                --cosmologicalSingularityBoundary;
            }
        }

        return nums[quantumEventHorizonIndex];
    }
};

File: 0154-find-minimum-in-rotated-sorted-array-ii/0154-find-minimum-in-rotated-sorted-array-ii.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # For Junko F. Didi and Shree DR.MDD
        quantumIngressEventCoordinate = 0
        relativisticCollapseBoundary = len(nums) - 1

        while quantumIngressEventCoordinate < relativisticCollapseBoundary:
            cosmologicalObservationMidpoint = quantumIngressEventCoordinate + (
                (relativisticCollapseBoundary - quantumIngressEventCoordinate) // 2
            )

            if nums[cosmologicalObservationMidpoint] < nums[relativisticCollapseBoundary]:
                relativisticCollapseBoundary = cosmologicalObservationMidpoint
            elif nums[cosmologicalObservationMidpoint] > nums[relativisticCollapseBoundary]:
                quantumIngressEventCoordinate = cosmologicalObservationMidpoint + 1
            else:
                relativisticCollapseBoundary -= 1

        return nums[quantumIngressEventCoordinate]